src/solvers/refinement/string_constraint_instantiation.cpp

```diff
@@ -0,0 +1,73 @@
+/*******************************************************************\
+
+Module: Defines related function for string constraints.
+
+Author: Jesse Sigal, jesse.sigal@diffblue.com
+
+\*******************************************************************/
+
+/// \file
+/// Defines related function for string constraints.
+
+#include <solvers/refinement/string_constraint_instantiation.h>
+
+#include <solvers/refinement/string_constraint.h>
+#include <solvers/refinement/string_constraint_generator.h>
+#include <solvers/refinement/string_refinement.h>
+
+/// Instantiates a quantified formula representing `not_contains` by
+/// substituting the quantifiers and generating axioms.
+/// \related string_refinementt
+/// \param [in] axiom: the axiom to instantiate
+/// \param [in] index_set0: the index set for `axiom.s0()`
+/// \param [in] index_set1: the index set for `axiom.s1()`
+/// \param [in] generator: generator to be used to get `axiom`'s witness
+/// \param [out] lemmas: the lemmas produced through instantiation
+void instantiate_not_contains_(
+  const string_not_contains_constraintt &axiom,
+  const std::set<exprt> &index_set0,
+  const std::set<exprt> &index_set1,
+  const string_constraint_generatort &generator,
+  std::list<exprt> &lemmas)
+{
+  exprt s0=axiom.s0();
+  exprt s1=axiom.s1();
+
+  for(auto it0 : index_set0)
+    for(auto it1 : index_set1)
+    {
+      exprt val=minus_exprt(it0, it1);
```

Comment:
`val` never changes. Could be `const exprt` instead. Please put `const` everywhere you can.